### PR TITLE
Set safe point at background

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve? 
Currently, we just set the service GC safe point when writing SST files, but not when checksumming.

If checksum time expires 5min (the service safe point lifetime) plus the TiDB GC lifetime, we might meet "GC life time is shorter than transaction duration" error.

### What is changed and how it works?
I moved the logic of setting GC safe point into another goroutine, which will set the service safe point every (1/5 service safe point lifetime).

This goroutine will be canceled by context.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
    Though this test can be automatized by fail points, if we did so, our ingratiation test time will grow to 15 mins+, which is hard to accept. 
    Firstly, we insert fail point at here:    
image
    Then, we run current master branch with fail point:
image
    After fixed, we can finish the backup:
image

### Release Note

 - Fix a bug that may cause failed to backup when dataset is too large and GC lifetime is too brief.

### CODE REVIEW

File: pkg/backup/safe_point.go [hunk 1]
```diff
@@ -52,3 +54,32 @@ func UpdateServiceSafePoint(ctx context.Context, pdClient pd.Client, ttl int64,
...
+// StartServiceSafePointKeeper will run UpdateServiceSafePoint periodicity
+// hence keeping service safepoint won't lose.
+func StartServiceSafePointKeeper(
+	ctx context.Context,
+	ttl int64,
+	pdClient pd.Client,
+	backupTS uint64,
+) {
+	// At least 1 second gap, or time.NewTicker will blame us.
+	gapSec := time.Duration(utils.MaxInt(int(ttl/5), 1))
```
Comment (kennytm) [inline pkg/backup/safe_point.go:67]: what's special about "5"?

btw please make this a true duration:

```go
gap := time.Duration(ttl) * time.Second / 5
if gap < time.Second {
    gap = time.Second
}
```

Comment (YuJuncen) [reply pkg/backup/safe_point.go:67]: The default ttl is 5 mins... So I just choice one of its factor.

Do we need so frequently reset it? The origin version reset it every time a range is fully backed up.

Comment (YuJuncen) [reply pkg/backup/safe_point.go:67]: (And every 5 seconds...)

Comment (kennytm) [reply pkg/backup/safe_point.go:67]: As long as it is reset before the TTL it is fine

**Comment by kennytm** [comment]
what if we just remove the loop inside BackupRange and move it entirely into StartServiceSafePointKeeper? @3pointer

**Comment by YuJuncen** [comment]
> what if we just remove the loop inside BackupRange and move it entirely into StartServiceSafePointKeeper? @3pointer

I think that would be a good idea 👍

File: pkg/backup/client.go [hunk 2]
```diff
@@ -439,44 +441,14 @@ func (bc *Client) BackupRanges(
...
-			}
-			if err != nil {
-				return nil, err
-			}
-		case <-t.C:
-		}
 	}
+
+	<-allFilesCollected
+	return allFiles, nil
```
Comment (kennytm) [inline pkg/backup/client.go:451]:
```suggestion
	select {
	case <-allFilesCollected:
		return allFiles, nil
	case <-ctx.Done():
		return nil, errors.Trace(ctx.Err())
	}
```

maybe combine with the `errCh` too, like if `errCh` returns `nil` treat it as success.

Comment (YuJuncen) [reply pkg/backup/client.go]: `errCh` is closed at this context, maybe there is no need to select on it.

File: pkg/backup/client.go [hunk 3]
```diff
@@ -447,8 +447,14 @@ func (bc *Client) BackupRanges(
...
 
-	<-allFilesCollected
-	return allFiles, nil
+	select {
+	case <-allFilesCollected:
+		return allFiles, nil
+	case <-ctx.Done():
+		return nil, errors.Trace(ctx.Err())
+	case err := <-errCh:
+		return nil, errors.Trace(err)
```
Comment (kennytm) [inline pkg/backup/client.go]: (you should remove the `for err := range errCh` loop above if handling this here)

**Pull request merged**